pkg/ucp/dataprovider/storageprovider.go

```diff
@@ -18,71 +18,119 @@ package dataprovider
 
 import (
 	"context"
-	"errors"
+	"fmt"
 	"sync"
 
 	"github.com/radius-project/radius/pkg/ucp/store"
-	"github.com/radius-project/radius/pkg/ucp/util"
 )
 
-var (
-	ErrUnsupportedStorageProvider = errors.New("unsupported storage provider")
-	ErrStorageNotFound            = errors.New("storage provider not found")
-)
+// DataStorageProvider acts as a factory for storage clients.
+//
+// Do not use construct this directly:
+//
+// - Use DataStorageProviderFromOptions instead for production use.
+// - Use DataStorageProviderFromMemory or DataStorageProviderFromClient for testing.
+type DataStorageProvider struct {
+	// options configures the settings of the storage provider.
+	options StorageProviderOptions
+
+	// factory is factory function to create a new storage client. Can be overridden for testing.
+	factory storageFactoryFunc
+
+	// init is used to guarantee single-initialization of the storage provider.
+	init sync.RWMutex
 
-var _ DataStorageProvider = (*storageProvider)(nil)
+	// result is result of invoking the factory (cached).
+	result result
+}
+
+type result struct {
+	client store.StorageClient
+	err    error
+}
 
-type storageProvider struct {
-	clients   map[string]store.StorageClient
-	clientsMu sync.RWMutex
-	options   StorageProviderOptions
+// DataStorageProviderFromOptions creates a new instance of the DataStorageProvider struct with the given options.
+//
+// This will used the known factory functions to instantiate the storage client.
+func DataStorageProviderFromOptions(options StorageProviderOptions) *DataStorageProvider {
+	return &DataStorageProvider{options: options}
 }
 
-// NewStorageProvider creates a new instance of the "storageProvider" struct with the given
-// "StorageProviderOptions" and returns it.
-func NewStorageProvider(opts StorageProviderOptions) DataStorageProvider {
-	return &storageProvider{
-		clients: map[string]store.StorageClient{},
-		options: opts,
+// DataStorageProviderFromMemory creates a new instance of the DataStorageProvider struct using the in-memory client.
+//
+// This will use the ephemeral in-memory storage client.
+func DataStorageProviderFromMemory() *DataStorageProvider {
+	return &DataStorageProvider{options: StorageProviderOptions{Provider: TypeInMemory}}
+}
+
+// DataStorageProviderFromClient creates a new instance of the DataStorageProvider struct with the given client.
+//
+// This will always return the given client and will not attempt to create a new one. This can be used for testing
+// with mocks.
+func DataStorageProviderFromClient(client store.StorageClient) *DataStorageProvider {
+	return &DataStorageProvider{result: result{client: client}}
+}
+
+// GetStorageClient returns a storage client for the given resource type.
```

Comment:
Comment needs update since resource type is not an input for client generated

Reply:
👍